rust-mode.el

```diff
@@ -267,18 +268,40 @@ to the function arguments.  When nil, `->' will be indented one level."
         ;; Rewind until the point no longer moves
         (setq continue (/= starting (point)))))))
 
-(defun rust-in-macro ()
+(defvar-local rust-macro-scopes nil)
+
+(defun rust-macro-scope (&optional start end)
   (save-excursion
-    (when (> (rust-paren-level) 0)
-      (backward-up-list)
-      (rust-rewind-irrelevant)
-      (or (rust-looking-back-macro)
-          (and (rust-looking-back-ident)
-               (save-excursion
-                 (backward-sexp)
-                 (rust-rewind-irrelevant)
-                 (rust-looking-back-str "macro_rules!")))
-          (rust-in-macro)))))
+    ;; need to special case macro_rules which has unique syntax
+    (let ((scope nil))
+      (goto-char
+       (or start (point-min)))
+      ;; if there is a start move back to the first top level
+      (when start
+        (let ((top (syntax-ppss-toplevel-pos (syntax-ppss))))
+          (when top
+            (goto-char top))))
+      (while (search-forward "!" end t)
+        (when (and (not (rust-in-str-or-cmnt))
+                   (skip-chars-forward " \t\n\r")
+                   (or
```

Comment:
nit: I usually see that as `(memq (char-after) ...)`

Reply:
Which?

Reply:
I mean to replace 
```(or  (eq ?\[ (char-after))  (eq ?\( (char-after))  (eq ?\{ (char-after)))```
 with 
```(memq (char-after) '(?\( ?\[ ?\{))```.

Reply:
Oh, yes, good call. Fixed.